system/lib/wasm_worker/library_wasm_worker.c

```diff
@@ -22,42 +24,52 @@ void __wasm_init_tls(void *memory);
 __attribute__((constructor(48)))
 static void emscripten_wasm_worker_main_thread_initialize() {
 	uintptr_t* sbrk_ptr = emscripten_get_sbrk_ptr();
-	assert((*sbrk_ptr & 15) == 0);
-	assert(__builtin_wasm_tls_align() <= 16);
+	*sbrk_ptr = ROUND_UP(*sbrk_ptr, __builtin_wasm_tls_align());
 	__wasm_init_tls((void*)*sbrk_ptr);
-	*sbrk_ptr += ROUND_UP(__builtin_wasm_tls_size(), 16);
+	*sbrk_ptr += ROUND_UP(__builtin_wasm_tls_size(), SBRK_ALIGN);
 }
 
 emscripten_wasm_worker_t emscripten_create_wasm_worker(void *stackPlusTLSAddress, size_t stackPlusTLSSize)
 {
 	assert(stackPlusTLSAddress != 0);
-	assert((uintptr_t)stackPlusTLSAddress % 16 == 0);
+	assert((uintptr_t)stackPlusTLSAddress % STACK_ALIGN == 0);
 	assert(stackPlusTLSSize > 0);
-	assert(stackPlusTLSSize % 16 == 0);
+	assert(stackPlusTLSSize % STACK_ALIGN == 0);
 
 	// Guard against a programming oopsie: The target Worker's stack cannot be part of the calling
 	// thread's stack.
 	assert(emscripten_stack_get_base() <= (uintptr_t)stackPlusTLSAddress || emscripten_stack_get_end() >= (uintptr_t)stackPlusTLSAddress + stackPlusTLSSize
 		&& "When creating a Wasm Worker, its stack should be located either in global data or on the heap, not on the calling thread's own stack!");
 
-	// We expect TLS area to need to be at most 16 bytes aligned
-	assert(__builtin_wasm_tls_align() == 0 || 16 % __builtin_wasm_tls_align() == 0);
-
 #ifndef NDEBUG
 	// The Worker's TLS area will be spliced off from the stack region, so the
 	// stack needs to be at least as large as the TLS region.
-	uint32_t tlsSize = ROUND_UP(__builtin_wasm_tls_size(), 16);
+	uint32_t tlsSize = ROUND_UP(__builtin_wasm_tls_size(), __builtin_wasm_tls_align());
 	assert(stackPlusTLSSize > tlsSize);
 #endif
+	// The TLS region lives at the start of the stack region (the lowest address
+	// of the stack).  Since the TLS data alignment may be larger than stack
+	// alignment, we may need to round up the lowest stack address to meet this
+	// requirment.
+	if (__builtin_wasm_tls_align()) {
+		uintptr_t tlsBase = (uintptr_t)stackPlusTLSAddress;
+		if (tlsBase % __builtin_wasm_tls_align() != 0) {
+			tlsBase = ROUND_UP(tlsBase, __builtin_wasm_tls_align());
+			size_t padding = tlsBase - (uintptr_t)stackPlusTLSAddress;
+			stackPlusTLSAddress = (void*)tlsBase;
+			stackPlusTLSSize -= padding;
+		}
+	}
```

Comment:
Can we remove the `if`s to make this simply read

```cpp
uintptr_t offset = (-stackPlusTLSAddress) & (__builtin_wasm_tls_align()-1);
stackPlusTLSSize -= offset;
stackPlusTLSAddress += offset;
```
or maybe if this gated to read
```cpp
if (__builtin_wasm_tls_align() > 16) {
  uintptr_t offset = (-stackPlusTLSAddress) & (__builtin_wasm_tls_align()-1);
  stackPlusTLSSize -= offset;
  stackPlusTLSAddress += offset;
}
```
we might help to get an optimizer fold this whole part away when it inlines a constant function `__builtin_wasm_tls_align() { return 16; }` to the caller here. (binaryen's wasm-opt should be able to do these types of optimizations, @kripken ?)

Reply:
Definitely, yes, such a tiny constant function will be inlined, then the result would be used to get rid of that comparison and possibly the if, based on the value.

Reply:
Looks like it worked.